keps/sig-api-machinery/4008-crd-ratcheting/README.md

```diff
@@ -0,0 +1,1036 @@
+
+1. CRD author gets complaints from users that they don't know their CRs have
+invalid values until the CR is processed by the controller and updates the status.
+2. CRD author notices OpenAPI schemas support adding value validations, and  happily annotates their schema.
+3. Since only validation was changed: no fields were removed or had their types changed, the author considers the change too minor to require a new
+CRD version.
+4. During testing users apply the updated CRD. Some users have pre-existing resources that passed validation before, and now fail.
+5. Controllers break since they can't update the Status of the custom resource: it fails validation.
+6. User's entire workflow is now broken until they repair all broken custom resources.
+7. User downgrades their CRD and complains to the CRD author
+8. CRD author abandons their attempt to improve useability of their CR.
+
+CRD authors tightening simple validation rules that their controllers tolerate
+shouldn't require a version bump.
+
+With the new feature the story becomes a success:
+
+1. Users complain to CRD author. 
+2. CRD author annotates schema
+3. CRD author does not bump version
+4. Users apply updated CRD schema
+5. Controllers can successfully update the status field of the CR
+6. User's workflow keeps humming
+7. User compliments cunning CRD author for their brilliance
+8. CRD author remembers to user value validations in their schemas more in the future.
+
+#### K8s Update Tightens CRD validation
+
+In this case the user has a CRD with a field with the following schema:
+
+```yaml
+type: string
+format: dns1123subdomain
+```
+
+Kubernetes does not recognize this as a format. Today, Kubernetes treats
+[unrecognized formats as a no-op](https://github.com/kubernetes/kubernetes/blob/f9f9e7177a328ae47bac6b823d57a99dfbfd310b/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/validation/formats.go#L53-L65) and ignores them. So users would be able
+to write fields not conforming to this format to custom resources.
+
+If Kubernetes wishes to support this format, users workflows will break upon
+upgrade until the offending resources and identified and repaired.
+
+With this CRD Validation Ratcheting feature, Kubernetes would be able to
+adopt new format strings safely without breaking user workflows.
+
+#### K8s Update Widens CRD Validation
+
+In this case the user has a CRD with a field with following schema:
+
+```yaml
+type: string
+format: byte
+```
+
+In k8s 1.27 and below, this format disallows empty strings due to an implementation detail. However, in native types like `Secret`, byte is disallowed. Is Kubernetes had an update to CRDs to allow empty string for `byte`, this widens the validation for those CRDS which use it.
+
+This is fine when upgrading to new versions if Kubernetes: all existing resources still pass validation.
+
+The user may still experience issues when downgrading, for example if they had unreletated issues with the new Kubernetes update:
```

Comment:
Yeah, basically, this case can be treated as an update to a tightened validation.

Reply:
added note to KEP